### DISCUSSION TITLE
Disassembler

### DISCUSSION THREAD

**Discussion by Ratstail91** [krgamestudios/Toy#112] [General]
@hiperiondev Let's move disassembler talk here, since it's basically a mini-project of it's own.

I reverted the #111 merge, because the default disassembly wasn't working correctly - in fact, it would hang in the default mode.

Do you think you could split the two modes apart, so they're not interleaved like they are now?

Edit: Also, since it's getting more complex, it might be a good idea to add in some simple tests for it too.

**Comment by hiperiondev** [comment]
I can see breaks standard format. I will correct this

**Comment by hiperiondev** [comment]
After the last option for disassembler maybe is time to rethink the interpreter

**Reply by Ratstail91** [reply]
How so?

**Reply by hiperiondev** [reply]
First, eliminate the nested structure of the program (that would practically be solved with the last option of the disassembler).
Modify stack handling by functions with a MARK structure (https://core.ac.uk/download/pdf/15449671.pdf)
With this reduce the Return only to recover the PC.
Create a more compact bytecode.

I'm going to do some testing on it.

**Reply by hiperiondev** [reply]
For example, an idea without much elaboration for the bytecode could be:

[0 | opcode (6 bits) | r ] r: return    <- with this, interpreting a return as an instruction is eliminated
[1 | lit type (5 bits) | len (3 bits) ] len : 0=next byte or 1, 2, 3 <- with this short literals of up to 3 elements already have the size specified, len can have other uses depending on the opcode

6 bits for opcode allow to 64 instructions, 5 bit for literal type to 32. Enough for future expansions
Retrieving the fields with a mask and a shift is one of the fastest things a microprocessor can do.

**Reply by Ratstail91** [reply]
I see. While you have my blessing to make these changes, I don't think they'll benefit the canonical branch (I'd rather keep it simple for others, and myself).

Keep me updated on what you do.

**Reply by hiperiondev** [reply]
Obviously this would be a new branch/project, it should not be mixed with what exists now. It shouldn't touch the compiler either, it would start with a code translator that should be able to run any code created in the current compiler.

**Comment by hiperiondev** [comment]
This is a keep alive packet :-)
I am with a lot of work, as soon as I get free I will return to the project.

**Reply by Ratstail91** [reply]
No worries, I've been distracted too lol
